### PR TITLE
RFE: Add plugin name field into plugin context use it to improve error reporting 

### PR DESCRIPTION
**Pull request by rpv-tomsk**
Superseedes #2011

### CODE REVIEW

**Comment by rpv-tomsk** [comment]
I'm in progress of work on this enhancement.

First commit, which originally was written by @octo, rebased to current master.

After it was applied, Collectd starts to log following output:

```
[2018-05-26 20:26:27] logfile plugin: plugin_load: plugin "logfile" successfully loaded.
[2018-05-26 20:26:27] logfile plugin: logfile: invalid loglevel [debug] defaulting to 'info'
[2018-05-26 20:26:27] csv plugin: plugin_load: plugin "csv" successfully loaded.
[2018-05-26 20:26:27] network plugin: plugin_load: plugin "network" successfully loaded.
[2018-05-26 20:26:27] routeros plugin: plugin_load: plugin "routeros" successfully loaded.
```

To restore correct logging new function `daemon_log` was added (second commit).
Propose to discuss draft of this.

@rubenk, can you please dig into this feature a bit, to discuss core code changes?

Next commits should change plugins code - almost every file will be touched, so I want this work to be merged shortly to avoid multiple rebases with conflicts resolve.

**Comment by rpv-tomsk** [comment]
Added `bind` plugin as example of cleanup.

Bind plugin, before unfication: different behaviour of `QTypes` and `ParseTime` options:

[2018-05-26 22:37:59] bind plugin: Using string value `falses' for boolean option `ParseTime' is deprecated and will be removed in future releases. Use unquoted true or false instead.
[2018-05-26 22:37:59] bind plugin: Cannot parse string value `falses' of the `ParseTime' option as a boolean value.
[2018-05-26 22:37:59] bind plugin: The `QTypes' option needs exactly one boolean argument.


And after:


[2018-05-26 22:45:06] bind plugin: Using string value `falses' for boolean option `ParseTime' is deprecated and will be removed in future releases. Use unquoted true or false instead.
[2018-05-26 22:45:06] bind plugin: Cannot parse string value `falses' of the `ParseTime' option as a boolean value.
[2018-05-26 22:45:06] bind plugin: Using string value `trues' for boolean option `QTypes' is deprecated and will be removed in future releases. Use unquoted true or false instead.
[2018-05-26 22:45:06] bind plugin: Cannot parse string value `trues' of the `QTypes' option as a boolean value.

File: src/daemon/plugin.c [hunk 1]
```diff
@@ -2234,8 +2239,35 @@ void plugin_log(int level, const char *format, ...) {
...
     le = le->next;
   }
+} /* void _daemon_log */
+
+void plugin_log(int level, const char *format, ...) {
+  char msg[1024] = "";
+
+  char const *name = plugin_get_ctx().name;
+  if (name != NULL)
+    snprintf(msg, sizeof(msg), "%s plugin: ", name);
```
Comment (rubenk) [inline src/daemon/plugin.c]: please check for error / truncation.

File: src/daemon/plugin.c [hunk 2]
```diff
@@ -2234,8 +2239,35 @@ void plugin_log(int level, const char *format, ...) {
...
+
+  va_list ap;
+  va_start(ap, format);
+  vsnprintf(msg + strlen(msg), sizeof(msg) - strlen(msg), format, ap);
```
Comment (rubenk) [inline src/daemon/plugin.c]: likewise

File: src/daemon/plugin.c [hunk 3]
```diff
@@ -2234,8 +2239,35 @@ void plugin_log(int level, const char *format, ...) {
...
+  msg[sizeof(msg) - 1] = 0;
```
Comment (rubenk) [inline src/daemon/plugin.c]: not strictly needed if the vnsprintf call succeeds.

File: src/daemon/plugin.c [hunk 4]
```diff
@@ -2234,8 +2239,35 @@ void plugin_log(int level, const char *format, ...) {
...
+  va_end(ap);
+
+  _daemon_log(level, msg);
 } /* void plugin_log */
 
+void daemon_log(int level, const char *format, ...) {
+  char msg[1024] = "";
```
Comment (rubenk) [inline src/daemon/plugin.c]: just a thought, but why limit the length to 1024 bytes?

File: src/daemon/plugin.c [hunk 5]
```diff
@@ -346,19 +346,22 @@ static void log_list_callbacks(llist_t **list, /* {{{ */
 static int create_register_callback(llist_t **list, /* {{{ */
                                     const char *name, void *callback,
                                     user_data_t const *ud) {
-  callback_func_t *cf;
 
-  cf = calloc(1, sizeof(*cf));
+  if ((name == NULL) || (callback == NULL))
```
Comment (rubenk) [inline src/daemon/plugin.c]: no need for the inner braces.

**Comment by rubenk** [comment]
Hi @rpv-tomsk. I left some minor remarks, but otherwise this looks like a great improvement, go for it!

**Comment by rpv-tomsk** [comment]
>go for it!

Ok, just as a last question. 
Maybe a softer implementation should go: change only `configfile.c` messages, without touching `INFO() / WARNINIG() / ERROR()` macro?

File: src/daemon/plugin.c (hunk 1 above)
Comment (rpv-tomsk) [reply src/daemon/plugin.c]: What should we do on error ?

File: src/daemon/plugin.c (hunk 4 above)
Comment (rpv-tomsk) [reply src/daemon/plugin.c]: >"640k ought to be enough for anybody" 

I don't know... Historical reasons..
)

File: src/daemon/plugin.c (hunk 1 above)
Comment (rubenk) [reply src/daemon/plugin.c]: either pass it up to the caller, or just log it.

File: src/daemon/plugin.c (hunk 4 above)
Comment (rubenk) [reply src/daemon/plugin.c]: would be nice to allocate it dynamically using alloca or vla's, stacks are large enough these days.

**Comment by rubenk** [comment]
> Maybe a softer implementation should go: change only configfile.c messages, without touching INFO() / WARNINIG() / ERROR() macro?

I'm fine with whatever you prefer.

**Comment by rpv-tomsk** [comment]
Closed in favor of already merged #2821.

**Pull request closed without merging**